timemachine/integrator.py

```diff
@@ -88,6 +88,75 @@ def step(self, x, v):
         return new_x, new_v
 
 
+class VelocityVerletIntegrator(Integrator):
+    def __init__(self, force_fxn, masses, dt):
+        """WARNING: `.step` makes 2x more calls to force_fxn per timestep than `.multiple_steps`"""
+        self.dt = dt
+        self.masses = masses[:, np.newaxis]  # TODO: cleaner way to handle (n_atoms,) vs. (n_atoms, 3) mismatch?
+        self.force_fxn = force_fxn
+
+    def step(self, x, v):
+        """WARNING: makes 2 calls to force_fxn per timestep -- prefer `.multiple_steps` in most cases"""
+        v_mid = v + (0.5 * self.dt / self.masses) * self.force_fxn(x)
+        x = x + self.dt * v_mid
+        v = v_mid + (0.5 * self.dt / self.masses) * self.force_fxn(x)
+
+        return x, v
+
+    def multiple_steps(self, x, v, n_steps=1000):
+        # note: intermediate timesteps are staggered
+        #    xs[0], vs[0] = x_0, v_0
+        #    xs[1], vs[1] = x_2, v_{1.5}
+        #    xs[2], vs[2] = x_3, v_{2.5}
+        #    ...
+        #    xs[T-1], vs[T-1] = x_T, v_{T-0.5}
+        #    xs[T],   vs[T]   = x_T, v_T
+
+        # note: reorders loop slightly to avoid ~n_steps extraneous calls to force_fxn
+
+        zs = [(x, v)]
+
+        # initialize traj
+        v = v + (0.5 * self.dt / self.masses) * self.force_fxn(x)
+        x = x + self.dt * v
+
+        # run n_steps-1 steps
+        for t in range(n_steps - 1):
+            v = v + (self.dt / self.masses) * self.force_fxn(x)
```

Comment:
<img width="375" alt="Screen Shot 2022-07-28 at 3 03 24 PM" src="https://user-images.githubusercontent.com/2280724/181617324-fddf8b5e-370f-4a49-85f9-c3f94af4d8ba.png">

This looks identical to the OpenMM implementation - you mentioned an inconsistency but I'm not seeing it?

Reply:
Just wanted to draw your attention to the inclusion of initialize and finalize half-steps.

Note this introduces an inconsistency between `.step` and `.multiple_steps` 

https://github.com/proteneer/timemachine/blob/a3d1a49fcbb8191c3d8ee1f96723c487775224be/timemachine/integrator.py#L107-L113

* `.step` will update both x and v by a full timestep -- this is inefficient (requires 2 force evaluations) and is only intended for reference
* `.multiple_steps` will save x and v "staggered by a half-step" for all intermediates (as I think you would get if you queried [OpenMM context](http://docs.openmm.org/7.6.0/userguide/theory/04_integrators.html#verletintegrator) every step). Only the initial `(xs[0], vs[0])` and final `(xs[T], vs[T])` frames returned by this function will actually be "on-step." Could clean this up to return `(xs[t], vs[t]) = (x_t, v_t)` for all `t` to avoid headaches / be self-consistent, but initially I just care about the first and last frames.

Reply:
thanks! this makes sense to me - we can deal with the half-step issues later and just document it now.